File: MPI/insertion_sort/insertion_sort_parallel_Pipeline.c

#include <stdio.h>
#include <stdlib.h>
#include "mpi.h"

#define tag 100  /* tag used in MPI_Send */

void rand_init_array(int array[], int n);
void insertion_sort(int array[], int n, int rank, int* smaller, int* received);
void fetch_results(int sorted_array[], int n, int rank, int* smaller);
void display_array(int array[], int n);

int main(int argc, char *argv[]){
    int rank, size;
    int received, smaller;
    MPI_Status status;
    int init_array[size], sorted_array[size];

    MPI_Init(&argc,&argv);
    MPI_Comm_size(MPI_COMM_WORLD,&size);
    MPI_Comm_rank(MPI_COMM_WORLD,&rank);
    
    if(!rank){
            rand_init_array(init_array, size);  /* Process 0: Initialize array */
            (void) printf("Initial array: ");
            display_array(init_array, size);
    }

    insertion_sort(init_array, size, rank, &smaller, &received);

    fetch_results(sorted_array, size, rank, &smaller);

    MPI_Finalize();

    return 0;
}

/*
 * Function:  insertion_sort 
 * --------------------
 * Sorts the array's elements (size in total)
 *
 *  array: the initial array
 *  n: number of elements in the array
 *  rank: the process that runs the function
 *  smaller: array's smallest element
 *  received: array's element that was last received
 *
 */

void insertion_sort(int array[], int n, int rank, int* smaller, int* received){
    MPI_Status status;
    int i;
    for(i=0; i<n-rank; ++i){

        if(!rank)
            *received = array[i];
        else
            MPI_Recv(received, 1, MPI_INT, rank-1, tag, MPI_COMM_WORLD, &status);

        if(!i)  /* if this is the first step */
            *smaller = *received; /* initializes smaller with received at step zero */
        else
            if(*received > *smaller)  /* compares smaller to received */
                MPI_Send(received, 1, MPI_INT, rank+1, tag, MPI_COMM_WORLD);   /* sends the larger number through */
        else{
            MPI_Send(smaller, 1, MPI_INT, rank+1, tag, MPI_COMM_WORLD);    /* sends the larger number through */
            *smaller = *received;
        }
    }
}

/*
 * Function:  fetch_results 
 * --------------------
 * Fetches elements from all processes (size in total)
 *
 *  sorted_array: the array with initial array's sorted elements 
 *  n: number of elements in the array
 *  rank: the process that runs the function
 *  smaller: array's smallest element
 *
 */

void fetch_results(int sorted_array[], int n, int rank, int* smaller){
    MPI_Status status;
    int i;
    if(!rank){  /* if is process 0 */
        sorted_array[0] = *smaller;
        for(i=1; i<n; ++i)
            MPI_Recv(&sorted_array[i], 1, MPI_INT, i, tag, MPI_COMM_WORLD, &status);
        (void) printf("\nSorted array: ");
        display_array(sorted_array, n);
   }
   else
        MPI_Send(smaller, 1, MPI_INT, 0, tag, MPI_COMM_WORLD); /* send the smaller element to process 0 */
}

/*
 * Function:  rand_init_array 
 * --------------------
 * Fills an integer array with random numbers
 *
 *  array: the array that will be filled with numbers
 *  n: number of elements in the array
 *
 */

void rand_init_array(int array[], int n){
    int i;    
    for (i=0; i<n; ++i)
        array[i] = (rand() % n*i) + 1;
}

/*
 * Function:  display_array 
 * --------------------
 * Prints an integer array to user
 *
 *  array: the array that will be printed
 *  n: number of elements in the array
 *
 */

void display_array(int array[], int n){
    (void) printf("[ ");
    int i;
    for (i=0; i<n; ++i)
        (void) printf("%d ", array[i]);
    (void) printf("]\n\n");
}

File: MPI/mpi4py/insertion_sort/insertion_sort_parallel_Pipeline.py

import numpy as np
from mpi4py import MPI

TAG = 200
ROOT = 0

def fetch_results(sorted_array, n, rank, smaller, comm):
    '''
    Function:  fetch_results 
    --------------------
    Fetches elements from all processes (size in total)

    sorted_array: the array with initial array's sorted elements 
    n: number of elements in the array
    rank: the process that runs the function
    smaller: array's smallest element
    '''
    status = MPI.Status()
    if(rank == ROOT):  # if is process 0
        sorted_array[0] = smaller
        for i in range (1, n):
            sorted_array[i] = comm.recv(source=i, tag=TAG, status=status)
    else:
        comm.send(obj=smaller, dest=ROOT, tag=TAG) # send the smaller element to process 0

def insertion_sort(array, n, rank, comm):
    '''
    Function:  insertion_sort 
    --------------------
    Sorts the array's elements (size in total)
    
    array: the initial array
    n: number of elements in the array
    rank: the process that runs the function
    smaller: array's smallest element
    received: array's element that was last received
    '''
    status = MPI.Status()
    for i in range(n-rank):
        if rank == ROOT:
            received = array[i]
        else:
            received = comm.recv(source=rank-1, tag=TAG, status=status)

        if not i:  # if this is the first step
            smaller = received; # initializes smaller with received at step zero
        else:
            if received > smaller:  # compares smaller to received
                comm.send(obj=received, dest=rank+1, tag=TAG)   # sends the larger number through
            else:
                comm.send(obj=smaller, dest=rank+1, tag=TAG)  # sends the larger number through
                smaller = received
    return smaller

if __name__ == '__main__':

    comm = MPI.COMM_WORLD
    size = comm.Get_size()
    rank = comm.Get_rank()

    a = np.empty(size, dtype=np.int)
    if rank == ROOT:
        lower, upper  = 0, size*4
        a = np.random.randint(low=lower, high=upper, size=size)
        print(f'\nInitial Array: {a}')
        begin = MPI.Wtime()
        
    smaller = insertion_sort(a, size, rank, comm)
    sorted_a = np.empty(size, dtype=np.int)
    fetch_results(sorted_array=sorted_a, n=size, rank=rank, smaller=smaller, comm=comm)

    if rank == ROOT:
        end = MPI.Wtime()
        print(f'\nSorted Array: {sorted_a}')
        print(f'\nTime elapsed for operation: {end-begin} seconds.\n')